Add a "X-Version" header that exposes the Fastly configuration version

kptdobe opened the pull request:
PR for https://github.com/adobe/helix-cli/issues/247

kptdobe commented:
Now how to make it clean: 

* if we agree on the approach, the only thing I do not like is to generate the X-Version line in the `strains.vcl` file.
* 2 options:
  1. create a new file for other future generated stuff 
  2. rename the `strains.vcl` to `dynamic.vcl` which contains everything that need to be generated  and re-format the code a little bit to aggregate the various dynamic parts in various functions.

I vote for option 2. Simple and easy.

@trieloff @tripodsan WDYT ?

trieloff commented:
2 sounds good. Here are a couple of things that would look good:
- VCL version number
- Git Commit SHA
- HLX CLI Version Number

trieloff commented on layouts/fastly/helix.vcl:
```diff
@@ -28,7 +28,7 @@ sub hlx_strain {
 
     # run custom strain resolution
-    include "strains.vcl";
+    include "dynamic.vcl";
```
I would probably split the two includes. Keep one `strains.vcl` and have an entirely new `dynamic.vcl`

kptdobe commented:
Added some few things:
- kept the `strains.vcl` file and added a `dynamic.vcl` file
- slightly changed the publish process to fix a bug: `helix.vcl` cannot be deployed if `strains.vcl` or `dynamic.vcl` does not exist (include is checked and fails)! This would block the really first deployment...
- `x-version` looks like this: `20 | 0.8.1-pre.4 | 46f590ef25ab5beb060d537f5fb2db91c1e61bec`
- format is `<fastly config version> | <hlx cli version> | <current folder version (project code)>`
- I changed to output X-Version all the time (no X-Debug required). I think this is really useful until things get stable and helps really being sure the content viewed in the browser is generated with latest code. We can change that back when things are rolling.

> VCL version number

What do you mean ? We do not have such a version number. But the cli version should give you a hint.

Still TODO: add the tests for all the changes

trieloff commented:
VCL version number is covered with `<fastly config version>`.

Not super thrilled about having the header in by default, as our headers tend to get chattier over time and then we have to consolidate again.

trieloff commented:
Maybe we could have a "weak debug" option, that we turn on when the user agent is `curl`, etc.

kptdobe commented:
Arff... not so simple! The `dynamic.vcl` is... dynamic thus cannot predict (record and replay) the POST request to upload the vcl.

kptdobe commented:
> Not super thrilled about having the header in by default, as our headers tend to get chattier over time and then we have to consolidate again.

@trieloff I clearly see that doing the same request with `X-Debug` and without triggers a new action call which means the 2 requests are cached independently. This also meaning we lose the information with which versions the non-debug request has been initially generated with when we want to debug a production site.
I am just saying we might need it with the first site(s) going on production for "debugging". Once this is stable or if this is useless, we can easily get rid of it. I created https://github.com/adobe/helix-cli/issues/264 to track the removal.

tripodsan commented:
maybe we should ask fastly to put the configuration version somewhere in their https://docs.fastly.com/vcl/variables/.

tripodsan commented:
message from fastly:

> Hi Tobias,
>
> I think the req.vcl variable is the best one for your purposes, below is a quick definition:
>
> req.vcl - STRING - Read Only - Available in vcl_recv
> The version of the customer VCL currently loaded. It will take the form [service_id].> [version_number]_[generation]-[hash].
>
> I've also created a fiddle for you to see what the output of this looks like:  https://fiddle.fastlydemo.net/fiddle/e7282e34

tripodsan commented:
I quickly tested it, and it works. for example, I get:

```
72jihuXMkAS9sNuyA7SFpL.20_7-47c7204417d30617822e1eb6543cab77
```

kptdobe commented:
Interesting, `req.vcl` is not documented...

Since @trieloff wanted to have more than the config version (cli version and current folder last commit sha), the x-version format is `<fastly config version> | <hlx cli version> | <current folder revision (project code)>`.
We anyway need to compute the `X-Version` dynamically to append the last 2. And since gathering the version is anyway the first thing the publish command is doing, I do not see the need to use this variable for now. If we would do everything in VCL, yes, but we compute everything in JS anyway.
